app/src/main/java/zapsolutions/zap/util/ExchangeRateUtil.java

```diff
@@ -0,0 +1,308 @@
+            JsonObjectRequest request;
+
+            switch (provider) {
+                case BLOCKCHAIN_INFO:
+                    request = getBlockchainInfoRequest();
+                    break;
+                case COINBASE:
+                    request = getCoinbaseRequest();
+                    break;
+                default:
+                    request = getBlockchainInfoRequest();
+            }
+
+            if (request != null) {
+                // Adding request to request queue
+                HttpClient.getInstance().addToRequestQueue(request, "rateRequest");
+                ZapLog.debug(LOG_TAG, "Exchange rate request initiated");
+            }
+        }
+    }
+
+
+    /**
+     * Creates a Request that fetches fiat exchange rate data from "blockchain.info".
+     * When executed this request saves the result in shared preferences and
+     * updates the currentCurrency of the MonetaryUtil Singleton.
+     *
+     * @return JsonObjectRequest
+     */
+    private JsonObjectRequest getBlockchainInfoRequest() {
+        JsonObjectRequest rateRequest = new JsonObjectRequest(Request.Method.GET, "https://blockchain.info/ticker", null,
+                response -> {
+                    ZapLog.debug(LOG_TAG, "Received exchange rates from blockchain.info");
+                    JSONObject responseRates = parseBlockchainInfoResponse(response);
+                    applyExchangeRatesAndSaveInPreferences(responseRates);
+                }, error -> ZapLog.debug(LOG_TAG, "Fetching exchange rates from blockchain.info failed"));
+
+        return rateRequest;
+    }
+
+    /**
+     * Creates a Request that fetches fiat exchange rate data from Coinbase.
+     * When executed this request saves the result in shared preferences and
+     * updates the currentCurrency of the MonetaryUtil Singleton.
+     *
+     * @return JsonObjectRequest
+     */
+    private JsonObjectRequest getCoinbaseRequest() {
+        JsonObjectRequest rateRequest = new JsonObjectRequest(Request.Method.GET, "https://api.coinbase.com/v2/exchange-rates?currency=BTC", null,
+                response -> {
+                    ZapLog.debug(LOG_TAG, "Received exchange rates from coinbase");
+                    JSONObject responseRates = parseCoinbaseResponse(response);
+                    applyExchangeRatesAndSaveInPreferences(removeNonFiat(responseRates));
+                }, error -> ZapLog.debug(LOG_TAG, "Fetching exchange rates from coinbase failed"));
+
+        return rateRequest;
+    }
+
+    /**
+     * This function parses a blockchain.info exchange rate response.
+     * All the response parser functions return a similar formatted JSON Object
+     * {"USD":{"rate":0.1231, "timestamp":...},"EUR":{...}}
+     *
+     * @param response a JSON response that comes from Blockchain.info
+     * @return
+     */
+    public JSONObject parseBlockchainInfoResponse(JSONObject response) {
```

Comment:
Maybe an easier way to parse the Json would be by using `Gson`. 
see e.g. https://stackoverflow.com/questions/38636254/how-to-convert-json-to-java-object-using-gson
But works as well as we do it now 👍